File: codewars/6kyu/array-diff/dduck-array-diff.py
# https://www.codewars.com/kata/523f5d21c841566fde000009

from unittest import TestCase


def array_diff(a, b):
    b_dict = dict.fromkeys(b)
    return list(filter(lambda key: key not in b_dict, a))


TestCase().assertEqual(array_diff([1, 2], [1]), [2], "a was [1,2], b was [1], expected [2]")
TestCase().assertEqual(array_diff([1, 2, 2], [1]), [2, 2], "a was [1,2,2], b was [1], expected [2,2]")
TestCase().assertEqual(array_diff([1, 2, 2], [2]), [1], "a was [1,2,2], b was [2], expected [1]")
TestCase().assertEqual(array_diff([1, 2, 2], []), [1, 2, 2], "a was [1,2,2], b was [], expected [1,2,2]")
TestCase().assertEqual(array_diff([], [1, 2]), [], "a was [], b was [1,2], expected []")

"""
best practice in codewars

def array_diff(a, b):
    return [x for x in a if x not in b]

"""

File: codewars/6kyu/array-diff/dduck-array-diff.js
// https://www.codewars.com/kata/523f5d21c841566fde000009

const array_diff = (a, b, bSet = new Set(b)) => a.filter((v) => !bSet.has(v));


console.log(array_diff([], [4, 5]), [], 'a was [], b was [4,5]');
console.log(array_diff([3, 4], [3]), [4], 'a was [3,4], b was [3]');
console.log(array_diff([1, 8, 2], []), [1, 8, 2], 'a was [1,8,2], b was []');
